src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -525,13 +525,67 @@ public static SpecialType GetSpecialTypeSafe(this TypeSymbol? type)
 
         public static bool IsAtLeastAsVisibleAs(this TypeSymbol type, Symbol sym, ref HashSet<DiagnosticInfo>? useSiteDiagnostics)
         {
-            HashSet<DiagnosticInfo>? localUseSiteDiagnostics = useSiteDiagnostics;
-            var result = type.VisitType((type1, symbol, unused) => IsTypeLessVisibleThan(type1, symbol, ref localUseSiteDiagnostics), sym,
+            var closure = IsAtLeastAsVisibleAsClosure.Get(useSiteDiagnostics);
+
+            var result = type.VisitType(closure.Func, sym,
                                         canDigThroughNullable: true); // System.Nullable is public
-            useSiteDiagnostics = localUseSiteDiagnostics;
+
+            useSiteDiagnostics = closure.Release();
             return result is null;
         }
 
+        // Type to reduce allocations of the Func<TypeSymbol, Symbol, bool, bool> closure in the above method
+        private class IsAtLeastAsVisibleAsClosure
```

Comment:
❗ If something like this is needed, use [`PooledDelegates`](https://github.com/dotnet/roslyn/blob/master/src/Dependencies/PooledObjects/PooledDelegates.cs)